lib/semian/sysv_shared_memory.rb

```diff
@@ -0,0 +1,75 @@
+module Semian
+  module SysVSharedMemory #:nodoc:
+    @type_size = {}
+    def self.sizeof(type)
+      size = (@type_size[type.to_sym] ||= (respond_to?(:_sizeof) ? _sizeof(type.to_sym) : 0))
+      raise TypeError.new("#{type} is not a valid C type") if size <= 0
+      size
+    end
+
+    def self.included(base)
+      def base.do_with_sync(*names)
+        names.each do |name|
+          new_name = "#{name}_inner".freeze.to_sym
+          alias_method new_name, name
+          private new_name
+          define_method(name) do |*args, &block|
+            synchronize do
+              method(new_name).call(*args, &block)
+            end
+          end
+        end
+      end
+    end
+
+    def semid
+      -1
+    end
+
+    def shmid
+      -1
+    end
+
+    def synchronize(&block)
+      if respond_to?(:_synchronize) && @using_shared_memory
+        return _synchronize(&block)
+      else
+        yield if block_given?
+      end
+    end
```

Comment:
Example of a function that could probably just be implemented in C and would benefit from the simplification of not doing run-time fallbacks.